src/librustc_mir/interpret/place.rs

```diff
@@ -410,6 +410,12 @@ where
                 stride * field
             }
             layout::FieldPlacement::Union(count) => {
+                // This is a narrow bug-fix for rust-lang/rust#69191: if we are
+                // trying to access absent field of uninhabited variant, then
+                // signal UB (but don't ICE the compiler).
+                if field >= count as u64 && base.layout.abi == layout::Abi::Uninhabited {
+                    throw_ub!(Unreachable);
```

Comment:
This is quite the hack... Miri really shouldn't need special checking for `layout::Abi::Uninhabited`. The right behavior for the uninhabited type ought to just fall out of the correct behavior for all kinds of types.

@eddyb also seems like the hack of using `Union` layout for uninhabited enums is having some bad repercussions here? But generally I am wondering how this can ever be reached; creating an uninhabited enum shiuld already be UB earlier...

@pnkfelix to make this hack slightly less hacky, is it sufficient to only do this when `base.layout.ty.is_enum()`?

Reply:
This is reached from const prop...

Reply:
I think we should just merge this and figure out how to do it correctly after it's merged, so we get rid of the beta regression immediately

Reply:
Oh no, this is const-prop again. :(
From [the backtrace](https://github.com/rust-lang/rust/issues/69191#issuecomment-587952118):

```
  14: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::operand_field
  15: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::eval_place_to_op
  16: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::eval_operand
  17: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::eval_rvalue_into_place
  18: <rustc_mir::transform::const_prop::ConstPropagator as rustc::mir::visit::MutVisitor>::visit_statement
  19: <rustc_mir::transform::const_prop::ConstProp as rustc_mir::transform::MirPass>::run_pass
```

So likely const-prop is running some unreachable code, and rather unsurprisingly that is tripping other parts of the compiler. Cc @wesleywiser 

Here's the pre-regression MIR:
```
    bb0: {
        StorageLive(_1);                 // bb0[0]: scope 0 at src/main.rs:6:9: 6:14
        StorageLive(_2);                 // bb0[1]: scope 0 at src/main.rs:6:40: 6:44
        _2 = const Scalar(<ZST>) : std::option::Option<UninhabitedUnivariant>; // bb0[2]: scope 0 at src/main.rs:6:40: 6:44
                                         // ty::Const
                                         // + ty: std::option::Option<UninhabitedUnivariant>
                                         // + val: Value(Scalar(<ZST>))
                                         // mir::Constant
                                         // + span: src/main.rs:6:40: 6:44
                                         // + literal: Const { ty: std::option::Option<UninhabitedUnivariant>, val: Value(Scalar(<ZST>)) }
        _1 = const std::option::Option::<UninhabitedUnivariant>::unwrap(move _2) -> bb1; // bb0[3]: scope 0 at src/main.rs:6:40: 6:53
                                         // ty::Const
                                         // + ty: fn(std::option::Option<UninhabitedUnivariant>) -> UninhabitedUnivariant {std::option::Option::<UninhabitedUnivariant>::unwrap}
                                         // + val: Value(Scalar(<ZST>))
                                         // mir::Constant
                                         // + span: src/main.rs:6:45: 6:51
                                         // + literal: Const { ty: fn(std::option::Option<UninhabitedUnivariant>) -> UninhabitedUnivariant {std::option::Option::<UninhabitedUnivariant>::unwrap}, val: Value(Scalar(<ZST>)) }
    }

    bb1: {
        StorageDead(_2);                 // bb1[0]: scope 0 at src/main.rs:6:52: 6:53
        StorageLive(_3);                 // bb1[1]: scope 1 at src/main.rs:8:41: 8:43
        _3 = move ((_1 as _Variant).0: Void); // bb1[2]: scope 1 at src/main.rs:8:41: 8:43
        StorageDead(_3);                 // bb1[3]: scope 1 at src/main.rs:9:5: 9:6
        StorageDead(_1);                 // bb1[4]: scope 0 at src/main.rs:10:1: 10:2
        return;                          // bb1[5]: scope 0 at src/main.rs:10:2: 10:2
    }
```
In `bb1`, mir building attempt to access `(_1 as _Variant).0`, and that variant has `Union` layout, and thus the ICE.

I feel like there is something incoherent in the compiler: either layout generation should actually add an appropriate amount of fields to these variants, or MIR building should not emit field accesses to fields that do not exist in the layout. But it seems like a reaosnable invariant to expect that any field mentioned statically *anywhere* in the code, is also actually existing in the layout.

@eddyb is it possible to, at leats, make the layout of such dead variants actually have the right number of fields (as declared in the source)?

Reply:
All right, I'll open an issue then about the layout/MIR incoherence.

Reply:
:upside_down_face: we can also solve the ICE by moving https://github.com/rust-lang/rust/blob/master/src/librustc_mir/interpret/operand.rs#L368-L371 to the beginning of the function, thus avoiding the problematic situation entirely

Reply:
Can we really? `op.layout.field(self, field)` will likely fail because `field` is not a valid field index.

Reply:
I opened an issue for further discussion: https://github.com/rust-lang/rust/issues/69763

Reply:
> @eddyb is it possible to, at least, make the layout of such dead variants actually have the right number of fields (as declared in the source)?

They do, I even showed you the code and brought up that `Result<!, !>` has `Union(1)` for both variants days ago.

We fixed this months ago. I don't understand why this is a problem again now.